src/lib/optimizer/optimizer.cpp

```diff
@@ -105,7 +105,7 @@ std::shared_ptr<Optimizer> Optimizer::create_default_optimizer() {
   // for now we want the PredicateReorderingRule to have the final say on predicate positions
   optimizer->add_rule(std::make_shared<PredicatePlacementRule>());
 
-  // Bring predicates into the desired order once the PredicateReorderingRule has positioned them as desired
+  // Bring predicates into the desired order once the PredicatePlacementRule has positioned them as desired
   optimizer->add_rule(std::make_shared<PredicateReorderingRule>());
```

Comment:
Not part of this PR, but why do we place the `IndexScanRule` behind the predicate reordering rule? @SvenLehmann : something to address?

Reply:
I don't really like either way with Hyrise's current functionality.

If we first run the PredicateReorderingRule, we might miss a better order.
If we, however, first run the IndexScanRule and then reorder the nodes, we run into problems as IndexScans are only supported for non-referenced segments. Thus, we would have to check after reordering, whether there is an IndexScan that is run on referenced segments and rewrite it to a TableScan again. And if we revert the choice of using an Index, this might result in different costs for the chain of predicates. So reordering would be necessary again.

There are two solutions coming to my mind right now.
1. We could fix the IndexScan to handle ReferenceSegments (no idea whether this is desirable or what the design decisions were). In general, there should not be any differences in feature support between the TableScan operator and the IndexScan operator. Maybe, it would be good enough for now to fall back to the TableScan in the case of reference segments, similar to the IndexJoin falling back to the NestedLoopJoin in certain situations.

2. Merge both optimizer rules, so that the predicate reordering is aware of the possibility that the first predicate might use an index. I don't really like this approach, though, as it mingles two independent optimization steps.

Reply:
Let's discuss this on Wednesday.